File: Wechat_chatroom_helper_android/app/src/main/java/com/zdy/project/wechat_chatroom_helper/ui/chatroomView/ChatroomRecyclerViewAdapter.java
package com.zdy.project.wechat_chatroom_helper.ui.chatroomView;

import android.content.Context;
import android.graphics.Canvas;
import android.graphics.Paint;
import android.graphics.drawable.BitmapDrawable;
import android.graphics.drawable.ShapeDrawable;
import android.graphics.drawable.shapes.Shape;
import android.support.v7.widget.RecyclerView;
import android.view.View;
import android.view.ViewGroup;

import com.zdy.project.wechat_chatroom_helper.HookLogic;
import com.zdy.project.wechat_chatroom_helper.model.MessageEntity;
import com.zdy.project.wechat_chatroom_helper.utils.ScreenUtils;

import java.util.ArrayList;

import de.robv.android.xposed.XposedBridge;
import de.robv.android.xposed.XposedHelpers;

import static com.zdy.project.wechat_chatroom_helper.Constants.Method_Message_Status_Bean;
import static com.zdy.project.wechat_chatroom_helper.Constants.Method_Message_True_Content;
import static com.zdy.project.wechat_chatroom_helper.Constants.Method_Message_True_Time;
import static com.zdy.project.wechat_chatroom_helper.Constants.Value_Message_Bean_NickName;
import static com.zdy.project.wechat_chatroom_helper.Constants.Value_Message_True_Content_Params;

/**
 * Created by Mr.Zdy on 2017/8/27.
 */

public class ChatRoomRecyclerViewAdapter extends RecyclerView.Adapter<ChatRoomViewHolder> {

    private Context mContext;

    private Object originAdapter;

    private ArrayList<Integer> muteListInAdapterPositions = new ArrayList<>();

    private OnDialogItemClickListener onDialogItemClickListener;

    private ArrayList data = new ArrayList();

    public void setData(ArrayList data) {
        this.data = data;
    }

    public void setOnDialogItemClickListener(OnDialogItemClickListener onDialogItemClickListener) {
        this.onDialogItemClickListener = onDialogItemClickListener;
    }

    public void setMuteListInAdapterPositions(ArrayList<Integer> muteListInAdapterPositions) {
        this.muteListInAdapterPositions = muteListInAdapterPositions;
    }

    public ArrayList<Integer> getMuteListInAdapterPositions() {
        return muteListInAdapterPositions;
    }

    ChatRoomRecyclerViewAdapter(Context context, Object originAdapter) {
        mContext = context;
        this.originAdapter = originAdapter;
    }

    @Override
    public ChatRoomViewHolder onCreateViewHolder(ViewGroup parent, int viewType) {
        return new ChatRoomViewHolder(ChatRoomViewHelper.getItemView(mContext));
    }

    public Object getObject(int position) {
        return data.get(position);
    }

    public ArrayList getData() {
        return data;
    }

    @Override
    public void onBindViewHolder(ChatRoomViewHolder holder, final int position) {
//        Object item = HookLogic.getMessageBeanForOriginIndex(originAdapter, muteListInAdapterPositions.get(position));

        Object item = getObject(position);
        MessageEntity entity = new MessageEntity(item);

        try {
            Object j = XposedHelpers.callMethod(originAdapter, Method_Message_Status_Bean, item);

            CharSequence content = (CharSequence) XposedHelpers.callMethod(originAdapter, Method_Message_True_Content,
                    item, ScreenUtils.dip2px(mContext, 13), XposedHelpers.getBooleanField(j, Value_Message_True_Content_Params));

            CharSequence time = (CharSequence) XposedHelpers.callMethod(originAdapter, Method_Message_True_Time, item);

            holder.nickname.setText((CharSequence) XposedHelpers.getObjectField(j,
                    Value_Message_Bean_NickName));
            holder.content.setText(content == null ? entity.field_digest : content);
            holder.time.setText(time);

            XposedBridge.log("content =" + content + ", field_digest = " + entity.toString());
        } catch (Exception e) {
            e.printStackTrace();
        }

        HookLogic.setAvatar(holder.avatar, entity.field_username);


        if (entity.field_unReadCount > 0)
            holder.unread.setBackground(new ShapeDrawable(new Shape() {
                @Override
                public void draw(Canvas canvas, Paint paint) {
                    int size = canvas.getWidth();

                    paint.setAntiAlias(true);
                    paint.setColor(0xFFFF0000);
                    paint.setStyle(Paint.Style.FILL_AND_STROKE);
                    canvas.drawCircle(size / 2, size / 2, size / 2, paint);
                }
            }));
        else
            holder.unread.setBackground(new BitmapDrawable(mContext.getResources()));

        holder.itemView.setBackground(ChatRoomViewHelper.getItemViewBackground());


        holder.itemView.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                onDialogItemClickListener.onItemClick(position);
            }
        });
    }

    @Override
    public int getItemCount() {
        return data.size();
    }

    public interface OnDialogItemClickListener {
        void onItemClick(int relativePosition);
    }
}

File: Wechat_chatroom_helper_android/app/src/main/kotlin/com/zdy/project/wechat_chatroom_helper/ui/wechat/chatroomView/ChatRoomRecyclerViewAdapter.kt
package com.zdy.project.wechat_chatroom_helper.ui.wechat.chatroomView

import android.content.Context
import android.graphics.Canvas
import android.graphics.Color
import android.graphics.Paint
import android.graphics.drawable.BitmapDrawable
import android.graphics.drawable.ShapeDrawable
import android.graphics.drawable.shapes.Shape
import android.support.v7.widget.RecyclerView
import android.view.ViewGroup
import com.zdy.project.wechat_chatroom_helper.HookLogic
import com.zdy.project.wechat_chatroom_helper.R
import com.zdy.project.wechat_chatroom_helper.model.ChatInfoModel
import utils.AppSaveInfoUtils
import java.util.*

/**
 * Created by Mr.Zdy on 2017/8/27.
 */

class ChatRoomRecyclerViewAdapter internal constructor(private val mContext: Context) : RecyclerView.Adapter<ChatRoomViewHolder>() {

    var muteListInAdapterPositions = ArrayList<Int>()

    private lateinit var onDialogItemClickListener: OnDialogItemClickListener

    var data = ArrayList<ChatInfoModel>()

    fun setOnDialogItemClickListener(onDialogItemClickListener: OnDialogItemClickListener) {
        this.onDialogItemClickListener = onDialogItemClickListener
    }

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): ChatRoomViewHolder {
        return ChatRoomViewHolder(ChatRoomViewHelper.getItemView(mContext))
    }

    private fun getObject(position: Int): ChatInfoModel {
        return data[position]
    }

    override fun onBindViewHolder(holder: ChatRoomViewHolder, position: Int) {

        val item = getObject(position)

        holder.nickname.text = item.nickname
        holder.content.text = item.content
        holder.time.text = item.time

        try {
            HookLogic.setAvatar(holder.avatar, item.avatarString)
        } catch (e: Throwable) {
            e.printStackTrace()
            holder.avatar.setImageResource(R.mipmap.ic_launcher)
        }

        if (item.unReadCount > 0)
            holder.unread.background = ShapeDrawable(object : Shape() {
                override fun draw(canvas: Canvas, paint: Paint) {
                    val size = (canvas.width/2).toFloat()

                    paint.isAntiAlias = true
                    paint.color = -0x10000
                    paint.style = Paint.Style.FILL_AND_STROKE
                    canvas.drawCircle(size, size, size, paint)
                }
            })
        else
            holder.unread.background = BitmapDrawable(mContext.resources)

        holder.itemView.background = ChatRoomViewHelper.getItemViewBackground(mContext)
        holder.itemView.setOnClickListener {
            try {
                onDialogItemClickListener.onItemClick(muteListInAdapterPositions[position])
            } catch (t: Throwable) {
                t.printStackTrace()
            }
        }
        holder.itemView.setOnLongClickListener {
            try {
                onDialogItemClickListener.onItemLongClick(muteListInAdapterPositions[position])
            } catch (t: Throwable) {
                t.printStackTrace()
            }
            return@setOnLongClickListener true
        }

        holder.nickname.setTextColor(Color.parseColor("#" + AppSaveInfoUtils.nicknameColorInfo()))
        holder.content.setTextColor(Color.parseColor("#" + AppSaveInfoUtils.contentColorInfo()))
        holder.time.setTextColor(Color.parseColor("#" + AppSaveInfoUtils.timeColorInfo()))
        holder.divider.setBackgroundColor(Color.parseColor("#" + AppSaveInfoUtils.dividerColorInfo()))
    }

    override fun getItemCount(): Int {
        return data.size
    }

    interface OnDialogItemClickListener {
        fun onItemClick(relativePosition: Int)

        fun onItemLongClick(relativePosition: Int)
    }


}
